src/com/carolinarollergirls/scoreboard/defaults/DefaultRulesetsModel.java

```diff
@@ -0,0 +1,339 @@
+    private void setCurrentRulesetRecurse(String i) {
+        Ruleset rs = getRuleset(i);
+        if (!rs.getId().equals(rootId)) {
+            setCurrentRulesetRecurse(rs.getParentRulesetId());
+        }
+        current.putAll(rs.getAll());
+        name = rs.getName();
+        id = rs.getId();
+    }
+
+    public Map<String, String> getAll() {
+        synchronized (coreLock) {
+            return Collections.unmodifiableMap(current);
+        }
+    }
+    public String get(String k) {
+        synchronized (coreLock) {
+            return current.get(k);
+        }
+    }
+    public boolean getBoolean(String k) {
+        synchronized (coreLock) {
+            return Boolean.parseBoolean(get(k));
+        }
+    }
+    public int getInt(String k) {
+        synchronized (coreLock) {
+            return Integer.parseInt(get(k));
+        }
+    }
+    public long getLong(String k) {
+        synchronized (coreLock) {
+            return Long.parseLong(get(k));
+        }
+    }
+    public void set(String k, String v) {
+        synchronized (coreLock) {
+            Rule r = rules.get(k);
+            if (r == null || !r.isValueValid(v)) {
+                return;
+            }
+            current.put(k, v);
+            scoreBoardChange(new ScoreBoardEvent(this, EVENT_CURRENT_RULESET, "", ""));
+        }
+    }
+
+    public Map<String, Rule> getRules() {
+        synchronized (coreLock) {
+            return Collections.unmodifiableMap(rules);
+        }
+    }
+
+    public Ruleset getRuleset(String id) {
+        return getRulesetModel(id);
+    }
+    public RulesetModel getRulesetModel(String id) {
+        synchronized (coreLock) {
+            RulesetModel r = rulesets.get(id);
+            if (r == null) {
+                r = rulesets.get(rootId);
+            }
+            return r;
+        }
+    }
+    public Map<String, Ruleset> getRulesets() {
+        synchronized (coreLock) {
+            return Collections.unmodifiableMap(new HashMap<String, Ruleset>(rulesets));
+        }
+    }
+    public RulesetModel addRuleset(String name, String parentId) {
+        return addRuleset(name, parentId, UUID.randomUUID().toString());
+    }
+    public RulesetModel addRuleset(String name, String parentId, String id) {
+        synchronized (coreLock) {
+            RulesetModel r = new DefaultRulesetModel(name, parentId, id);
+            rulesets.put(id, r);
+            r.addScoreBoardListener(this);
+            scoreBoardChange(new ScoreBoardEvent(r, EVENT_RULESET, r, null));
+            return r;
+        }
+    }
+    public void removeRuleset(String id) {
+        synchronized (coreLock) {
+            if (id.equals(rootId)) {
+                return;
+            }
+            Ruleset r = rulesets.get(id);
+            if (r == null) {
+                return;
+            }
+            requestBatchStart();
+            // Point any rulesets with the deleted one as their parent
+            // to their grandparent.
```

Comment:
This should also copy any non-overwritten settings to the child, so the child isn't modified.

Reply:
If there is a setting that is overridden in the deleted ruleset but not overwritten again in the reparented child, this setting will change for the child. I would consider this unexpected behaviour.

Reply:
That's all handled in the UI code. The rules are only fully rendered in the JS, the WS only contains the overrides that a given ruleset is making.

Reply:
I think understand what you mean now, let me take a look at that.

Reply:
The UI doesn't do that either. Pressing the "Delete" button calls:
`WS.Command("RemoveRuleset", o);`
And the WS command only calls this function.
So overrides that are set in the deleted ruleset are not copied over to its children, changing them unless they happen to reoverride the rule.

Reply:
That's done.